File: impl/go/extract.go
package twill

// Candidate extraction from arbitrary text (SPEC §13.4).

func extractIsWhitespace(c byte) bool {
	return c == ' ' || c == '\t' || c == '\n' || c == '\r' || c == '\f'
}

func extractIsQuote(c byte) bool { return c == '"' || c == '\'' || c == '`' }

func extractIsStartBoundary(c byte) bool {
	return extractIsWhitespace(c) || extractIsQuote(c) || c == '.' || c == '}' || c == '>'
}

func extractIsEndBoundary(c byte) bool {
	return extractIsWhitespace(c) || extractIsQuote(c) || c == ']' || c == '{' || c == '=' || c == '\\' || c == '<'
}

func extractIsLetter(c byte) bool { return c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z' }

func extractIsNameChar(c byte) bool {
	return extractIsLetter(c) || isDigit(c) || c == '_' || c == '-'
}

// at returns the byte at i, or 0 past the end.
func at(text string, i int) byte {
	if i < 0 || i >= len(text) {
		return 0
	}
	return text[i]
}

// consumeBalanced consumes a balanced bracket group starting at i (the
// opener). It returns the index after the closer, or -1 when unbalanced or
// interrupted by a line break.
func consumeBalanced(text string, i int, open, close byte) int {
	depth := 0
	for j := i; j < len(text); j++ {
		c := text[j]
		if c == '\n' {
			return -1
		}
		if c == '\\' {
			j++
			continue
		}
		if c == open {
			depth++
		} else if c == close {
			depth--
			if depth == 0 {
				return j + 1
			}
		}
	}
	return -1
}

// consumeBracketSuffix consumes `[...]` or `(...)` at i.
func consumeBracketSuffix(text string, i int) int {
	switch at(text, i) {
	case '[':
		return consumeBalanced(text, i, '[', ']')
	case '(':
		return consumeBalanced(text, i, '(', ')')
	}
	return -1
}

// consumeModifier consumes `/modifier` at i (the slash) and returns the end.
func consumeModifier(text string, i int) int {
	c := at(text, i+1)
	if c == '[' || c == '(' {
		end := consumeBracketSuffix(text, i+1)
		if end == -1 {
			return i
		}
		return end
	}
	j := i + 1
	for j < len(text) {
		d := text[j]
		if extractIsNameChar(d) || d == '.' || d == '%' {
			j++
		} else {
			break
		}
	}
	if j == i+1 {
		return i
	}
	last := text[j-1]
	if last == '-' || last == '_' {
		return i
	}
	return j
}

// consumeVariant consumes a variant at i and returns the index after it
// (before the `:`), or -1.
func consumeVariant(text string, i int) int {
	c := at(text, i)
	if c == '[' {
		return consumeBalanced(text, i, '[', ']')
	}
	if !extractIsLetter(c) && c != '@' {
		return -1
	}
	j := i + 1
	for j < len(text) {
		d := text[j]
		if extractIsNameChar(d) {
			j++
			continue
		}
		if (d == '[' || d == '(') && text[j-1] == '-' {
			end := consumeBracketSuffix(text, j)
			if end == -1 {
				return -1
			}
			j = end
			continue
		}
		break
	}
	if j == i+1 && c == '@' {
		return -1
	}
	last := text[j-1]
	if last == '-' || last == '_' {
		return -1
	}
	if at(text, j) == '/' {
		j = consumeModifier(text, j)
	}
	return j
}

// consumeUtility consumes a utility at i and returns the index after it, or -1.
func consumeUtility(text string, i int) int {
	j := i
	if at(text, j) == '!' {
		j++
	}
	c := at(text, j)

	if c == '[' {
		end := consumeBalanced(text, j, '[', ']')
		if end == -1 {
			return -1
		}
		found := false
		for k := j; k < end; k++ {
			if text[k] == ':' {
				found = true
				break
			}
		}
		if !found {
			return -1
		}
		j = end
	} else {
		if c == '-' {
			next := at(text, j+1)
			if !extractIsLetter(next) && !isDigit(next) {
				return -1
			}
			j++
		} else if !extractIsLetter(c) && c != '@' {
			return -1
		}
		j++
		for j < len(text) {
			d := text[j]
			if extractIsNameChar(d) || d == '%' {
				j++
				continue
			}
			if d == '.' {
				if isDigit(text[j-1]) && isDigit(at(text, j+1)) {
					j++
					continue
				}
				break
			}
			if (d == '[' || d == '(') && text[j-1] == '-' {
				end := consumeBracketSuffix(text, j)
				if end == -1 {
					return -1
				}
				j = end
				continue
			}
			break
		}
		last := text[j-1]
		if last == '-' || last == '_' {
			return -1
		}
	}

	if at(text, j) == '/' {
		j = consumeModifier(text, j)
	}
	if at(text, j) == '!' {
		j++
	}
	return j
}

// consumeCandidate consumes a full candidate `(variant ":")* utility` at i.
func consumeCandidate(text string, i int) int {
	j := i
	for {
		end := consumeVariant(text, j)
		if end != -1 && at(text, end) == ':' {
			j = end + 1
			continue
		}
		break
	}
	return consumeUtility(text, j)
}

// consumeVariable consumes a `--name` custom property reference at i.
func consumeVariable(text string, i int) int {
	j := i + 2
	for j < len(text) && extractIsNameChar(text[j]) {
		j++
	}
	if j == i+2 {
		return -1
	}
	return j
}

// candidateSet is an insertion-ordered set of candidates.
type candidateSet struct {
	seen map[string]bool
	list []string
}

func newCandidateSet() *candidateSet { return &candidateSet{seen: map[string]bool{}} }

func (s *candidateSet) add(c string) bool {
	if s.seen[c] {
		return false
	}
	s.seen[c] = true
	s.list = append(s.list, c)
	return true
}

func (s *candidateSet) has(c string) bool { return s.seen[c] }

func (s *candidateSet) len() int { return len(s.list) }

func (s *candidateSet) slice() []string { return append([]string{}, s.list...) }

// ExtractCandidates extracts every candidate-like span from text (SPEC
// §13.4) in order of first appearance. Recall is favored over precision:
// unknown candidates are harmless.
func ExtractCandidates(text string) []string {
	set := newCandidateSet()
	extractInto(text, set)
	return set.list
}

func extractInto(text string, into *candidateSet) {
	length := len(text)
	i := 0
	for i < length {
		c := text[i]
		if i > 0 && !extractIsStartBoundary(text[i-1]) {
			i++
			continue
		}

		end := -1
		if c == '-' && at(text, i+1) == '-' {
			end = consumeVariable(text, i)
		} else if extractIsLetter(c) || c == '@' || c == '-' || c == '!' || c == '[' {
			end = consumeCandidate(text, i)
		}

		if end != -1 && end > i && (end == length || extractIsEndBoundary(text[end])) {
			into.add(text[i:end])
			i = end
			continue
		}

		// Skip to the next boundary so that substrings are never emitted.
		i++
		for i < length && !extractIsStartBoundary(text[i-1]) {
			i++
		}
	}
}

File: impl/ocaml/src/extract.ml
(* Candidate extraction from arbitrary text (SPEC §13.4). *)

let is_whitespace c = c = ' ' || c = '\t' || c = '\n' || c = '\r' || c = '\012'
let is_quote c = c = '"' || c = '\'' || c = '`'
let is_start_boundary c = is_whitespace c || is_quote c || c = '.' || c = '}' || c = '>'
let is_end_boundary c = is_whitespace c || is_quote c || c = ']' || c = '{' || c = '=' || c = '\\' || c = '<'
let is_letter = Utils.is_letter
let is_digit = Utils.is_digit
let is_name_char c = is_letter c || is_digit c || c = '_' || c = '-'

(* The byte at [i], or NUL past the end. *)
let at text i = if i < 0 || i >= String.length text then '\000' else text.[i]

(* Consumes a balanced bracket group starting at [i] (the opener). Returns
   the index after the closer, or -1 when unbalanced or interrupted by a
   line break. *)
let consume_balanced text i op cl =
  let n = String.length text in
  let rec go j depth =
    if j >= n then -1
    else
      let c = text.[j] in
      if c = '\n' then -1
      else if c = '\\' then go (j + 2) depth
      else if c = op then go (j + 1) (depth + 1)
      else if c = cl then if depth - 1 = 0 then j + 1 else go (j + 1) (depth - 1)
      else go (j + 1) depth
  in
  go i 0

let consume_bracket_suffix text i =
  match at text i with '[' -> consume_balanced text i '[' ']' | '(' -> consume_balanced text i '(' ')' | _ -> -1

(* Consumes `/modifier` at [i] (the slash) and returns the end. *)
let consume_modifier text i =
  let c = at text (i + 1) in
  if c = '[' || c = '(' then
    let e = consume_bracket_suffix text (i + 1) in
    if e = -1 then i else e
  else begin
    let n = String.length text in
    let j = ref (i + 1) in
    while !j < n && (is_name_char text.[!j] || text.[!j] = '.' || text.[!j] = '%') do incr j done;
    if !j = i + 1 then i
    else
      let last = text.[!j - 1] in
      if last = '-' || last = '_' then i else !j
  end

(* Consumes a variant at [i] and returns the index after it (before the
   `:`), or -1. *)
let consume_variant text i =
  let c = at text i in
  if c = '[' then consume_balanced text i '[' ']'
  else if (not (is_letter c)) && c <> '@' then -1
  else begin
    let n = String.length text in
    let j = ref (i + 1) in
    let failed = ref false in
    let continue = ref true in
    while !continue && (not !failed) && !j < n do
      let d = text.[!j] in
      if is_name_char d then incr j
      else if (d = '[' || d = '(') && text.[!j - 1] = '-' then begin
        let e = consume_bracket_suffix text !j in
        if e = -1 then failed := true else j := e
      end
      else continue := false
    done;
    if !failed then -1
    else if !j = i + 1 && c = '@' then -1
    else
      let last = text.[!j - 1] in
      if last = '-' || last = '_' then -1 else if at text !j = '/' then consume_modifier text !j else !j
  end

(* Consumes a utility at [i] and returns the index after it, or -1. *)
let consume_utility text i =
  let n = String.length text in
  let j = ref i in
  if at text !j = '!' then incr j;
  let c = at text !j in
  let ok =
    if c = '[' then begin
      let e = consume_balanced text !j '[' ']' in
      if e = -1 then false
      else
        let found = ref false in
        for k = !j to e - 1 do if text.[k] = ':' then found := true done;
        if not !found then false else (j := e; true)
    end
    else begin
      let start_ok =
        if c = '-' then begin
          let next = at text (!j + 1) in
          if (not (is_letter next)) && not (is_digit next) then false else (incr j; true)
        end
        else is_letter c || c = '@'
      in
      if not start_ok then false
      else begin
        incr j;
        let failed = ref false and continue = ref true in
        while !continue && (not !failed) && !j < n do
          let d = text.[!j] in
          if is_name_char d || d = '%' then incr j
          else if d = '.' then
            if is_digit text.[!j - 1] && is_digit (at text (!j + 1)) then incr j else continue := false
          else if (d = '[' || d = '(') && text.[!j - 1] = '-' then begin
            let e = consume_bracket_suffix text !j in
            if e = -1 then failed := true else j := e
          end
          else continue := false
        done;
        if !failed then false
        else
          let last = text.[!j - 1] in
          not (last = '-' || last = '_')
      end
    end
  in
  if not ok then -1
  else begin
    if at text !j = '/' then j := consume_modifier text !j;
    if at text !j = '!' then incr j;
    !j
  end

(* Consumes a full candidate `(variant ":")* utility` at [i]. *)
let consume_candidate text i =
  let rec go j =
    let e = consume_variant text j in
    if e <> -1 && at text e = ':' then go (e + 1) else j
  in
  consume_utility text (go i)

(* Consumes a `--name` custom property reference at [i]. *)
let consume_variable text i =
  let n = String.length text in
  let j = ref (i + 2) in
  while !j < n && is_name_char text.[!j] do incr j done;
  if !j = i + 2 then -1 else !j

(* An insertion-ordered set of candidates. *)
type candidate_set = { seen : (string, unit) Hashtbl.t; mutable list_rev : string list; mutable size : int }

let new_candidate_set () = { seen = Hashtbl.create 256; list_rev = []; size = 0 }

let add set c =
  if Hashtbl.mem set.seen c then false
  else begin
    Hashtbl.replace set.seen c ();
    set.list_rev <- c :: set.list_rev;
    set.size <- set.size + 1;
    true
  end

let has set c = Hashtbl.mem set.seen c
let to_list set = List.rev set.list_rev

let extract_into text into =
  let length = String.length text in
  let i = ref 0 in
  while !i < length do
    let c = text.[!i] in
    if !i > 0 && not (is_start_boundary text.[!i - 1]) then incr i
    else begin
      let e =
        if c = '-' && at text (!i + 1) = '-' then consume_variable text !i
        else if is_letter c || c = '@' || c = '-' || c = '!' || c = '[' then consume_candidate text !i
        else -1
      in
      if e <> -1 && e > !i && (e = length || is_end_boundary text.[e]) then begin
        ignore (add into (String.sub text !i (e - !i)));
        i := e
      end
      else begin
        (* Skip to the next boundary so that substrings are never emitted. *)
        incr i;
        while !i < length && not (is_start_boundary text.[!i - 1]) do incr i done
      end
    end
  done

(* Extracts every candidate-like span from text in order of first
   appearance. Recall is favored over precision: unknown candidates are
   harmless. *)
let extract_candidates text =
  let set = new_candidate_set () in
  extract_into text set;
  to_list set
